### PR TITLE
feat: use Halo2 compatible dependencies

### PR DESCRIPTION
**Pull request by DrPeterVanNostrand**

### CODE REVIEW

File: src/bls.rs [hunk 1]
```diff
@@ -1,15 +1,18 @@
...
-        G2Compressed, G2Prepared, G2Uncompressed, G1 as G1Projective, G2 as G2Projective,
-    },
-    Compress, Engine, PairingCurveAffine,
-};
+use ff::PrimeField;
+use pairing::Engine;
+
+// Associates a `pairing::Engine` with its G1 base field; used to generate GPU field arithmetic for
+// that field.
+pub trait G1BaseField: Engine {
```
Comment (dignifiedquire) [inline src/bls.rs:11]: Move this out, as a regular export, maybe calling it sth like `ExtendedEngine` and define the base fields for both `G1` and `G2`

File: src/gpu/fft.rs [hunk 2]
```diff
@@ -26,7 +30,7 @@ where
 
 impl<E> FFTKernel<E>
 where
-    E: Engine,
+    E: Engine + G1BaseField,
```
Comment (dignifiedquire) [inline src/gpu/fft.rs:33]: only specify `E: G1BaseField`, as that already requires `Engine`

**Comment by DrPeterVanNostrand** [comment]
Note that this PR removes the ci jobs `test_pairing_gpu_x86_64-unknown-linux-gnu` and `test_pairing_x86_64-unknown-linux-gnu` and those github checks will never pass. I'm not sure how to remove those github checks and/or I may not have permission in CircleCI.

File: Cargo.toml [hunk 3]
```diff
@@ -34,37 +34,30 @@ itertools = "0.9.0"
 bincode = "1.3.1"
 serde = { version = "1.0", features = ["derive"] }
 sha2 = "0.9"
-
-# blst feature
-blstrs = { version = "0.3", optional = true }
-
-# pairing feature
-paired = { version = "0.22.0", optional = true }
+blstrs = { git = "https://github.com/filecoin-project/blstrs", branch = "halo" }
```
Comment (DrPeterVanNostrand) [inline Cargo.toml:37]: This should be changed before merging.

File: Cargo.toml [hunk 4]
```diff
@@ -34,37 +34,30 @@ itertools = "0.9.0"
...
+pairing = "0.20"
 
 # gpu feature 
 rust-gpu-tools = { version = "0.4.0", optional = true }
-ff-cl-gen = { version = "0.3.0", optional = true }
+ff-cl-gen = { git = "https://github.com/filecoin-project/ff-cl-gen", branch = "halo", optional = true }
```
Comment (DrPeterVanNostrand) [inline Cargo.toml:42]: This should be changed before merging.

**Review by DrPeterVanNostrand — COMMENTED** [review]
Cargo.toml's should be updated before merging.

File: src/groth16/tests/dummy_engine.rs [hunk 5]
```diff
@@ -21,224 +29,298 @@ impl fmt::Display for Fr {
...
     }
+}
 
-    fn sub_assign(&mut self, other: &Self) {
-        self.0 = ((MODULUS_R + self.0) - other.0) % MODULUS_R;
+impl Sub<&Fr> for &Fr {
+    type Output = Fr;
+
+    fn sub(self, rhs: &Fr) -> Fr {
+        self + -rhs
```
Comment (vmx) [inline src/groth16/tests/dummy_engine.rs:91]: `+ -` might be needed here, but it reads funny.

File: src/multiexp.rs [hunk 6]
```diff
@@ -194,54 +197,87 @@ impl DensityTracker {
     }
 }
 
+// Right shift the repr of a field element by `n` bits.
```
Comment (vmx) [inline src/multiexp.rs:200]: I haven't reviewed this part in-depth, if needed, let me know.

File: src/multiexp.rs [hunk 7]
```diff
@@ -265,44 +301,50 @@ where
...
+    E: gpu::GpuEngine,
+    E: Engine<Fr = G::Scalar>,
+    <<E as Engine>::Fr as PrimeField>::Repr: Copy + Send + Sync,
     S: SourceBuilder<G>,
 {
     if let Some(ref mut kern) = kern {
-        if let Ok(p) = kern.with(|k: &mut gpu::MultiexpKernel<G::Engine>| {
-            let mut exps = vec![exponents[0]; exponents.len()];
+        if let Ok(p) = kern.with(|k: &mut gpu::MultiexpKernel<E>| {
+            // let mut exps = vec![exponents[0]; exponents.len()];
```
Comment (vmx) [inline src/multiexp.rs:343]: left-over.

File: tests/groth16_aggregation.rs [hunk 8]
```diff
@@ -81,7 +83,7 @@ impl<'a, E: Engine> Circuit<E> for MimcDemo<'a, E> {
             // tmp = (xL + Ci)^2
             let tmp_value = xl_value.map(|mut e| {
                 e.add_assign(&self.constants[i]);
-                e.square();
+                e = e.square();
```
Comment (vmx) [inline tests/groth16_aggregation.rs:86]: `e.square()` can be returned directly.

File: tests/mimc.rs [hunk 9]
```diff
@@ -89,7 +92,7 @@ impl<'a, E: Engine> Circuit<E> for MimcDemo<'a, E> {
             // tmp = (xL + Ci)^2
             let tmp_value = xl_value.map(|mut e| {
                 e.add_assign(&self.constants[i]);
-                e.square();
+                e = e.square();
```
Comment (vmx) [inline tests/mimc.rs:95]: `e.square()` can be returned directly.

**Review by vmx — COMMENTED** [review]
Lots of `mul_assign()` where replaced by `+`, but not all of them (e.g. in `src/groth16/aggregate/poly.rs`). Not sure if that's a problem or not.

Some calculations changed a bit, but I'd expect that the test capture it if they were wrong now.

Lot's of comments, but all the them are rather style wise and nothing really important. So feel free to ignore those.

File: src/groth16/mapped_params.rs [hunk 10]
```diff
@@ -143,31 +149,35 @@ pub fn read_g1<E: Engine>(
...
+    let affine: E::G1Affine = {
+        let affine_opt = if checked {
+            E::G1Affine::from_uncompressed(&repr)
+        } else {
+            E::G1Affine::from_uncompressed_unchecked(&repr)
+        };
+
+        Option::from(affine_opt)
+            .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidData, "not on curve"))
+    }?;
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply src/groth16/mapped_params.rs:165]: honestly not super happy with either, so gonna leave it as is for now

File: src/groth16/tests/dummy_engine.rs (hunk 5 above)
Comment (dignifiedquire) [reply src/groth16/tests/dummy_engine.rs:91]: yeah I think it is needed, definitely looks funny

File: Cargo.toml [hunk 11]
```diff
@@ -16,55 +16,50 @@ edition = "2018"
...
-paired = { version = "0.22.0", optional = true }
+blstrs = { git = "https://github.com/filecoin-project/blstrs", branch = "master" }
+pairing = "0.21"
+yastl = "0.1.2"
 
 # gpu feature 
 rust-gpu-tools = { version = "0.4.0", optional = true }
-ff-cl-gen = { version = "0.3.0", optional = true }
+ec-gpu = { git = "https://github.com/filecoin-project/ec-gpu", branch = "master", optional = true }
+ec-gpu-gen = { git = "https://github.com/filecoin-project/ec-gpu", branch = "master", optional = true }
```
Comment (cryptonemo) [inline Cargo.toml:44]: Just to clarify, is ff-cl-gen going to be released before this is merged, or are we sitting at `master` while things are in flux?

**Review by cryptonemo — APPROVED** [review]
Looks good.  Exciting :+1:

File: Cargo.toml (hunk 11 above)
Comment (dignifiedquire) [reply Cargo.toml:44]: The plan is to sit on `master` until at least the cuda things are also merged

**Pull request merged**